src/backend/distributed/planner/merge_planner.c

```diff
@@ -13,92 +13,315 @@
 											  skipOuterVars);
-static DeferredErrorMessage * InsertDistributionColumnMatchesSource(Oid targetRelationId,
-																	Query *query);
+static DeferredErrorMessage * DeferErrorIfRoutableMergeNotSupported(Query *query,
+																	List *rangeTableList,
+																	PlannerRestrictionContext
+																	*
+																	plannerRestrictionContext);
 static DeferredErrorMessage * MergeQualAndTargetListFunctionsSupported(Oid
 																	   resultRelationId,
 																	   FromExpr *joinTree,
 																	   Node *quals,
 																	   List *targetList,
 																	   CmdType commandType);
 
+static DistributedPlan * CreateRouterMergePlan(Oid targetRelationId, Query *originalQuery,
+											   Query *query,
+											   List *rangeTableList,
+											   PlannerRestrictionContext *
+											   plannerRestrictionContext);
+static void ErrorIfRepartitionMergeNotSupported(Oid targetRelationId, Query *mergeQuery,
+												Query *sourceQuery);
+static void ConvertSourceRTEIntoSubquery(Query *mergeQuery,
+										 RangeTblEntry *sourceRte,
+										 PlannerRestrictionContext *
+										 plannerRestrictionContext);
+static void ConvertCteRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte);
+static void ConvertRelationRTEIntoSubquery(Query *mergeQuery,
+										   RangeTblEntry *sourceRte,
+										   PlannerRestrictionContext *
+										   plannerRestrictionContext);
+static void ErrorIfUnsupportedCTEs(Query *query);
+static void ContainsUnsupportedCTEs(Query *query);
+static bool MergeQueryCTEWalker(Node *node, void *context);
+static DistributedPlan * CreateRepartitionMergePlan(Oid targetRelationId, uint64 planId,
+													Query *originalQuery,
+													Query *query,
+													PlannerRestrictionContext *
+													plannerRestrictionContext,
+													ParamListInfo boundParams);
+static char * MergeCommandResultIdPrefix(uint64 planId);
+
+#endif
+
 
 /*
- * ErrorIfDistTablesNotColocated Checks to see if
- *
- *   - There are a minimum of two distributed tables (source and a target).
- *   - All the distributed tables are indeed colocated.
- *
- * If any of the conditions are not met, it raises an exception.
+ * CreateMergePlan
+ * 1) Check for conditions that are not supported in MERGE command.
+ * 2) Try to create a pushable plan
+ *    - Check for conditions suitable for a routable plan, if not found,
+ *      raise deferred error
+ * 3) Try to create repartition and redistribution plan
+ *    - Check for conditions that prevent repartition strategy, if found,
+ *      raise an exception and quit.
  */
-static DeferredErrorMessage *
-ErrorIfDistTablesNotColocated(Query *parse, List *distTablesList,
-							  PlannerRestrictionContext *
-							  plannerRestrictionContext)
+DistributedPlan *
+CreateMergePlan(uint64 planId, Query *originalQuery, Query *query,
```

Comment:
There are certain crashes / errors:
```SQL
 MERGE INTO target_dist t
        USING (SELECT count(DISTINCT key) as key FROM source_dist GROUP BY value) s
        ON t.key = s.key
WHEN NOT MATCHED THEN INSERT VALUES(s.key, 1) WHEN MATCHED THEN UPDATE SET value =150;
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
The connection to the server was lost. Attempting reset: Failed.
 @:-!>
```

or 
```SQL
 MERGE INTO target_dist t
        USING (SELECT count(DISTINCT value) as key FROM source_dist GROUP BY key) s
        ON t.key = s.key
WHEN NOT MATCHED THEN INSERT VALUES(s.key, 1) WHEN MATCHED THEN UPDATE SET value =150;
ERROR:  row field count is 2, expected 1
CONTEXT:  while executing command on localhost:9701
```

example schema:
```SQL
 \d source_dist
             Table "public.source_dist"
┌────────┬────────┬───────────┬──────────┬─────────┐
│ Column │  Type  │ Collation │ Nullable │ Default │
├────────┼────────┼───────────┼──────────┼─────────┤
│ key    │ bigint │           │          │         │
│ value  │ text   │           │          │         │
└────────┴────────┴───────────┴──────────┴─────────┘

[local] onderkalaci@onderkalaci:9700-14871=# \d target_dist
             Table "public.target_dist"
┌────────┬────────┬───────────┬──────────┬─────────┐
│ Column │  Type  │ Collation │ Nullable │ Default │
├────────┼────────┼───────────┼──────────┼─────────┤
│ key    │ bigint │           │          │         │
│ value  │ text   │           │          │         │
└────────┴────────┴───────────┴──────────┴─────────┘
```

Reply:
Something is amiss, do you mind sharing the sql script. Please print DEBUG too for the crash, since sub-query is not pushable, we must go through the re-partition.

```
MERGE INTO citus_target t
USING (SELECT count(DISTINCT sid) as sid FROM citus_source GROUP BY val) s
ON t.tid = s.sid
WHEN NOT MATCHED THEN INSERT VALUES(s.sid, 1) WHEN MATCHED THEN UPDATE SET val =150;
psql:crash:20: DEBUG:  Distributed tables are not co-located, try repartitioning
psql:crash:20: DEBUG:  For MERGE command, all the distributed tables must be colocated
psql:crash:20: DEBUG:  Creating MERGE repartition plan
psql:crash:20: DEBUG:  Router planner cannot handle multi-shard select queries
psql:crash:20: DEBUG:  Collect source query results on coordinator
psql:crash:20: ERROR:  MERGE INSERT must refer a source column for distribution column

```

Reply:
```
CREATE TABLE citus_target(tid int, val int);
CREATE TABLE citus_source(sid int, val int, const int);

SELECT create_distributed_table('citus_target', 'tid');
SELECT create_distributed_table('citus_source', 'sid', colocate_with=>'none');
--SELECT create_distributed_table('citus_source', 'sid', colocate_with=>'citus_target');

set citus.log_remote_commands to true;
SET client_min_messages TO DEBUG2;

MERGE INTO citus_target t
USING (SELECT count(DISTINCT sid) as sid FROM citus_source GROUP BY val) s
ON t.tid = s.sid
WHEN NOT MATCHED THEN INSERT VALUES(s.sid, 1) WHEN MATCHED THEN UPDATE SET val =150;

```

Reply:
Ok, sorry for the confusion. Yes, it seems I did the following code change while testing:

```
diff --git a/src/backend/distributed/executor/merge_executor.c b/src/backend/distributed/executor/merge_executor.c
index 379884d09..4b55c3a9c 100644
--- a/src/backend/distributed/executor/merge_executor.c
+++ b/src/backend/distributed/executor/merge_executor.c
@@ -364,9 +364,9 @@ sourceResultPartitionColumnIndex(Query *mergeQuery, List *sourceTargetList,
        int partitionColumnIndex = -1;
        if (sourceJoinColumn)
        {
-               ErrorIfInsertNotMatchTargetDistributionColumn(targetRelation->relationId,
-                                                                                                         mergeQuery,
-                                                                                                         sourceJoinColumn);
+               //ErrorIfInsertNotMatchTargetDistributionColumn(targetRelation->relationId,
+               //                                                                                        mergeQuery,
+               //                                                                                        sourceJoinColumn);

                /*
                 * Find the index in the source list that will be utilized
```


Because it seems otherwise, we are preventing a very important use-case. As a result, users do not really have the flexibility to use arbitrary subqueries as the source.

Given that we repartition the results, why do we still require the source distribution column? Isn't the main point of repartitioning that we do not require the distribution column of the source and align/push the results of the source with the target distribution key?

Reply:
>>Given that we repartition the results, why do we still require the source distribution column? 

We don't need it, it's a bug/limitation where we were checking only for `Expr `of type `Var`, but here we get `FuncExpr` (an implicit cast function `int4`) as side-effect of the aggregate function.

Reply:
The unexpected field counts are coming from the resjunk=true columns, it's fixed now.